src/table.rs

```diff
@@ -230,161 +286,112 @@ impl Table {
-    fn new(ingredient: IngredientIndex) -> Self {
-        Self {
-            ingredient,
-            allocated: Default::default(),
-            allocation_lock: Default::default(),
-            data: Box::new([const { UnsafeCell::new(MaybeUninit::uninit()) }; PAGE_LEN]),
-        }
+impl<'p, T: Slot> PageView<'p, T> {
+    fn page_data(&self) -> &[PageDataEntry<T>] {
+        let len = self.0.allocated.load(Ordering::Acquire);
+        // SAFETY: `len` is the initialized length of the page
+        unsafe { slice::from_raw_parts(self.0.data.cast::<PageDataEntry<T>>().as_ptr(), len) }
     }
 
-    fn check_bounds(&self, slot: SlotIndex) {
-        let len = self.allocated.load(Ordering::Acquire);
-        assert!(
-            slot.0 < len,
-            "out of bounds access `{slot:?}` (maximum slot `{len}`)"
-        );
-    }
-
-    /// Returns a reference to the given slot.
-    ///
-    /// # Panics
-    ///
-    /// If slot is out of bounds
-    pub(crate) fn get(&self, slot: SlotIndex) -> &T {
-        self.check_bounds(slot);
-        unsafe { (*self.data[slot.0].get()).assume_init_ref() }
-    }
-
-    /// Returns a reference to the given slot.
-    ///
-    /// # Panics
-    ///
-    /// If slot is out of bounds
-    pub(crate) fn get_mut(&mut self, slot: SlotIndex) -> &mut T {
-        self.check_bounds(slot);
-        unsafe { (*self.data[slot.0].get()).assume_init_mut() }
-    }
-
-    pub(crate) fn slots(&self) -> impl Iterator<Item = &T> {
-        let len = self.allocated.load(std::sync::atomic::Ordering::Acquire);
-        let mut idx = 0;
-
-        std::iter::from_fn(move || {
-            let ret = if idx < len {
-                let value = self.get(crate::table::SlotIndex::new(idx));
-                Some(value)
-            } else {
-                None
-            };
-            idx += 1;
-            ret
-        })
-    }
-
-    /// Returns a raw pointer to the given slot.
-    ///
-    /// # Panics
-    ///
-    /// If slot is out of bounds
-    ///
-    /// # Safety
-    ///
-    /// Safe to call, but reads/writes through this pointer must be coordinated
-    /// properly with calls to [`get`](`Self::get`) and [`get_mut`](`Self::get_mut`).
-    pub(crate) fn get_raw(&self, slot: SlotIndex) -> *mut T {
-        self.check_bounds(slot);
-        self.data[slot.0].get().cast()
+    fn data(&self) -> &'p [T] {
+        let len = self.0.allocated.load(Ordering::Acquire);
+        // SAFETY: `len` is the initialized length of the page
+        unsafe { slice::from_raw_parts(self.0.data.cast::<T>().as_ptr(), len) }
     }
 
     pub(crate) fn allocate<V>(&self, page: PageIndex, value: V) -> Result<Id, V>
     where
         V: FnOnce(Id) -> T,
     {
-        let guard = self.allocation_lock.lock();
-        let index = self.allocated.load(Ordering::Acquire);
-        if index == PAGE_LEN {
+        let _guard = self.0.allocation_lock.lock();
+        let index = self.0.allocated.load(Ordering::Acquire);
```

Comment:
Given we only ever modify `allocated` in this function while the `allocation_lock` is held, could we use `Relaxed` ordering for the load here ? I believe so as the `lock` call should already incur a happens before relationship for the participating threads right?